cube/swiss/source/devices/kunaigc/deviceHandler-KunaiGC.c

```diff
@@ -0,0 +1,250 @@
+	  0ULL,     // fileBase (u64)
+	  0,        // offset
+	  0,        // size
+	  IS_DIR,
+	  0
+	};
+	
+device_info initial_KunaiGC_info = {
+	0x200000,
+	0x200000,
+	false
+};
+
+static void readLFSInfo(void) {
+	initial_KunaiGC_info.totalSpace = (cfg.block_size * cfg.block_count);
+	initial_KunaiGC_info.freeSpace = initial_KunaiGC_info.totalSpace - (lfs_fs_size(&lfs) * cfg.block_size);
+}
+
+device_info* deviceHandler_KunaiGC_info(file_handle* file) {
+	return &initial_KunaiGC_info;
+}
+
+s32 deviceHandler_KunaiGC_makeDir(file_handle* file) {	
+	return lfs_mkdir(&lfs, file->name);
+}
+
+s32 deviceHandler_KunaiGC_readDir(file_handle* ffile, file_handle** dir, u32 type) {	
+	lfs_dir_t lfsdir;
+	struct lfs_info info;
+
+	int err = lfs_dir_open(&lfs, &lfsdir, ffile->name);
+	
+	if (err < 0) {
+		return -1;
+	}
+	
+	// Set everything up to read
+	int num_entries = 1, i = 1;
+	*dir = calloc(num_entries, sizeof(file_handle));
+	concat_path((*dir)[0].name, ffile->name, "..");
+	(*dir)[0].fileAttrib = IS_SPECIAL;
+	
+	// Read each entry of the directory
+	while (lfs_dir_read(&lfs, &lfsdir, &info) > 0) {
+		errno = 0;		
+
+		if(!strcmp(info.name, ".") || !strcmp(info.name, "..")) {
+			continue;
+		}
+		// Do we want this one?
+		if((type == -1 || ((info.type == LFS_TYPE_DIR) ? (type==IS_DIR) : (type==IS_FILE)))) {
+			struct lfs_info fstat;
+
+			if(info.type == LFS_TYPE_REG) {
+				if(!checkExtension(info.name)) continue;
+			}
+			// Make sure we have room for this one
+			if(i == num_entries){
+				++num_entries;
+				*dir = reallocarray(*dir, num_entries, sizeof(file_handle));
+			}
+			memset(&(*dir)[i], 0, sizeof(file_handle));
+			if(concat_path((*dir)[i].name, ffile->name, info.name) < PATHNAME_MAX
+				&& !lfs_stat(&lfs, (*dir)[i].name, &fstat) && fstat.size <= UINT32_MAX) {
+				(*dir)[i].size       = fstat.size;
+				(*dir)[i].fileAttrib = (fstat.type == LFS_TYPE_DIR) ? IS_DIR : IS_FILE;
+				++i;
+			}
+		}
+	};
+	
+	lfs_dir_close(&lfs, &lfsdir);
+	
+	return i;
+}
+
+s64 deviceHandler_KunaiGC_seekFile(file_handle* file, s64 where, u32 type) {
+	if(type == DEVICE_HANDLER_SEEK_SET) file->offset = where;
+	else if(type == DEVICE_HANDLER_SEEK_CUR) file->offset = file->offset + where;
+	else if(type == DEVICE_HANDLER_SEEK_END) file->offset = file->size + where;
+	return file->offset;
+}
+
+s32 deviceHandler_KunaiGC_readFile(file_handle* file, void* buffer, u32 length) {
+	if (!file->fp) {
+		if (lfs_file_open(&lfs, &lfs_file, file->name, LFS_O_RDONLY) < 0) return -1;
+		file->fp = &lfs_file;		
+	}
+	if(file->size <= 0) {
+		lfs_file_seek(&lfs, (lfs_file_t*)file->fp, 0, LFS_SEEK_END);
+		file->size = lfs_file_tell(&lfs, (lfs_file_t*)file->fp);
+	}
+	lfs_file_seek(&lfs, (lfs_file_t*)file->fp, file->offset, LFS_SEEK_SET);
+	size_t bytes_read = lfs_file_read(&lfs, (lfs_file_t*)file->fp, buffer, length);
+	file->offset = lfs_file_tell(&lfs, (lfs_file_t*)file->fp);
+		
+	return bytes_read;
+}
+
+// Assumes a single call to write a file.
```

Comment:
This comment isn't true because this device isn't considered for bulk file operations here: https://github.com/emukidid/swiss-gc/blob/6a06ad9d9e36cb7c1daec2b5dd0e59fa0e864a71/cube/swiss/source/swiss.c#L1723

@Extrems we should probably add a "Bulk read/write" attribute to the DEVICEHANDLER_INTERFACE?

Reply:
Woops - this slipped. The comment is irrelevant, I probably need to remove it. I basically copied Qoob Device, so its probably a left over.
In fact every file can be written in chunks, as I'm first seeking to the proper position within the file and then writing it.